### PR TITLE
Blog01 stanning

### PR DESCRIPTION
**Pull request by shmax**
Hey guys, I was feeling nostalgic about the old days of battling with the baseline errors file, so I thought I'd nibble away at it a bit. In this PR I focus on the stanning errors emitted by the blog-01 example.

### CODE REVIEW

File: examples/01-blog/Blog/Type/CommentType.php [hunk 1]
```diff
@@ -19,9 +18,27 @@ public function __construct()
...
-                'isAnonymous' => Types::boolean(),
+                'author' => [
+                    'type' => Types::user(),
+                    'resolve' => function (Comment $comment): ?User {
+                        if ($comment->isAnonymous) {
+                            return null;
+                        }
+
+                        return DataSource::findUser($comment->authorId);
+                    },
```
Comment (spawnia) [inline examples/01-blog/Blog/Type/CommentType.php:29]:
```suggestion
                    'resolve' => fn (Comment $comment): ?User => $comment->isAnonymous
                        ? null
                        : DataSource::findUser($comment->authorId),
```

File: examples/01-blog/Blog/Type/CommentType.php [hunk 2]
```diff
@@ -19,9 +18,27 @@ public function __construct()
...
+                ],
+                'parent' => [
+                    'type' => Types::comment(),
+                    'resolve' => function (Comment $comment): ?Comment {
+                        if ($comment->parentId !== null) {
+                            return DataSource::findComment($comment->parentId);
+                        }
+
+                        return null;
+                    },
```
Comment (spawnia) [inline examples/01-blog/Blog/Type/CommentType.php:39]:
```suggestion
                    'resolve' => fn (Comment $comment): ?Comment => $comment->parentId !== null
                        ? DataSource::findComment($comment->parentId)
                        : null,
```

File: examples/01-blog/Blog/Type/CommentType.php [hunk 3]
```diff
@@ -19,9 +18,27 @@ public function __construct()
...
+                ],
+                'isAnonymous' => ['type' => Types::boolean(), 'resolve' => static fn (Comment $comment): bool => $comment->isAnonymous],
```
Comment (spawnia) [inline examples/01-blog/Blog/Type/CommentType.php:41]: Can we just omit `resolve` here? The default resolver should do this.

File: examples/01-blog/Blog/Type/CommentType.php [hunk 4]
```diff
@@ -31,56 +48,23 @@ public function __construct()
...
+                        $args += ['after' => null];
+
+                        return DataSource::findReplies($comment->id, $args['limit'], $args['after']);
+                    },
+                ],
+                'totalReplyCount' => [
+                    'type' => Types::int(),
+                    'resolve' => function (Comment $comment): int {
+                        return DataSource::countReplies($comment->id);
+                    },
```
Comment (spawnia) [inline examples/01-blog/Blog/Type/CommentType.php:61]:
```suggestion
                    'resolve' => fn (Comment $comment): int => DataSource::countReplies($comment->id),
```

File: examples/01-blog/Blog/Type/Field/HtmlField.php [hunk 5]
```diff
@@ -24,8 +28,8 @@ public static function build(string $objectKey): array
                 ],
                 'maxLength' => Types::int(),
             ],
-            'resolve' => static function ($object, $args) use ($objectKey) {
-                $html = $object->{$objectKey};
+            'resolve' => static function ($value, $args) use ($config) {
```
Comment (spawnia) [inline examples/01-blog/Blog/Type/Field/HtmlField.php:31]:
```suggestion
            'resolve' => static function ($rootValue, array $args) use ($config): ?string {
```

File: examples/01-blog/Blog/Type/Field/HtmlField.php [hunk 6]
```diff
@@ -24,8 +28,8 @@ public static function build(string $objectKey): array
...
+                $html = $config['resolve']($value, $args);
```
Comment (spawnia) [inline examples/01-blog/Blog/Type/Field/HtmlField.php:32]: I would extract `$config['resolve']` outside of this callback - slight performance boost to not perform the array lookup multiple times

File: examples/01-blog/Blog/Type/StoryType.php [hunk 7]
```diff
@@ -20,10 +18,13 @@ public function __construct()
         parent::__construct([
             'name' => 'Story',
             'fields' => static fn (): array => [
-                'id' => Types::id(),
-                'author' => Types::user(),
-                'mentions' => new ListOfType(Types::mention()),
-                'totalCommentCount' => Types::int(),
+                'id' => ['type' => Types::id(), 'resolve' => static fn (Story $story): int => $story->id],
```
Comment (spawnia) [inline examples/01-blog/Blog/Type/StoryType.php:21]: should also work with the default resolver

File: examples/01-blog/Blog/Type/StoryType.php [hunk 8]
```diff
@@ -46,109 +54,41 @@ public function __construct()
...
+                'likedBy' => ['type' => new ListOfType(Types::user()), 'resolve' => static fn (Story $story) => DataSource::findLikes($story->id, 10)],
+                'affordances' => [
+                    'type' => new ListOfType(Types::storyAffordances()),
+                    'resolve' => function (Story $story, array $args, AppContext $context): array {
+                        /** @var array<int, string> $affordances */
+                        $affordances = [];
+
+                        $isViewer = $context->viewer === DataSource::findUser($story->authorId);
+                        if ($isViewer) {
+                            $affordances[] = Enum\StoryAffordancesType::EDIT;
```
Comment (spawnia) [inline examples/01-blog/Blog/Type/StoryType.php:68]: namespace can be simplified

File: examples/01-blog/Blog/Type/StoryType.php [hunk 9]
```diff
@@ -46,109 +54,41 @@ public function __construct()
...
+                            $affordances[] = Enum\StoryAffordancesType::DELETE;
+                        }
+
+                        $isLiked = DataSource::isLikedBy($story->id, $context->viewer->id);
+                        if ($isLiked) {
+                            $affordances[] = Enum\StoryAffordancesType::UNLIKE;
+                        } else {
+                            $affordances[] = Enum\StoryAffordancesType::LIKE;
+                        }
```
Comment (spawnia) [inline examples/01-blog/Blog/Type/StoryType.php:78]: ternary

File: examples/01-blog/Blog/Type/UserType.php [hunk 10]
```diff
@@ -19,22 +18,34 @@ public function __construct()
...
             'fields' => static fn (): array => [
-                'id' => Types::id(),
-                'email' => Types::email(),
+                'id' => [
+                    'type' => Types::id(),
+                    'resolve' => static fn (User $user): int => $user->id,
+                ],
+                'email' => [
+                    'type' => Types::email(),
+                    'resolve' => static fn (User $user): string => $user->email,
```
Comment (spawnia) [inline examples/01-blog/Blog/Type/UserType.php:27]: default resolver?

**Review by spawnia — COMMENTED** [review]
Nice, I like providing a higher quality example. I made some proposals to polish this even more.

File: examples/01-blog/Blog/Type/CommentType.php (hunk 1 above)
Comment (shmax) [reply examples/01-blog/Blog/Type/CommentType.php]: 999bff3b

File: examples/01-blog/Blog/Type/CommentType.php (hunk 2 above)
Comment (shmax) [reply examples/01-blog/Blog/Type/CommentType.php]: ed6c4e5c09cf0573c0cc30bac60b1cbc3f758369

File: examples/01-blog/Blog/Type/CommentType.php (hunk 3 above)
Comment (shmax) [reply examples/01-blog/Blog/Type/CommentType.php:33]: Yes! I forgot that was a thing, thank you 6060c965

File: examples/01-blog/Blog/Type/Field/HtmlField.php (hunk 5 above)
Comment (shmax) [reply examples/01-blog/Blog/Type/Field/HtmlField.php]: 42d0e041

File: examples/01-blog/Blog/Type/StoryType.php (hunk 7 above)
Comment (shmax) [reply examples/01-blog/Blog/Type/StoryType.php]: 342db408

File: examples/01-blog/Blog/Type/StoryType.php (hunk 8 above)
Comment (shmax) [reply examples/01-blog/Blog/Type/StoryType.php]: 13b7472c

File: examples/01-blog/Blog/Type/StoryType.php (hunk 9 above)
Comment (shmax) [reply examples/01-blog/Blog/Type/StoryType.php]: 3c6242f7

File: examples/01-blog/Blog/Type/Field/HtmlField.php (hunk 6 above)
Comment (shmax) [reply examples/01-blog/Blog/Type/Field/HtmlField.php]: Is this what you mean? 942370f7

**Comment by shmax** [comment]
@spawnia great review, thanks for being awesome. Let me know if I missed anything.

File: examples/01-blog/Blog/Type/CommentType.php [hunk 11]
```diff
@@ -31,56 +40,21 @@ public function __construct()
                             'defaultValue' => 5,
                         ],
                     ],
+                    'resolve' => function (Comment $comment, array $args): array {
+                        $args += ['after' => null];
+
+                        return DataSource::findReplies($comment->id, $args['limit'], $args['after']);
+                    },
```
Comment (spawnia) [inline examples/01-blog/Blog/Type/CommentType.php:47]:
```suggestion
                    'resolve' => fn (Comment $comment, array $args): array => DataSource::findReplies($comment->id, $args['limit'], $args['after'] ?? null),
```

File: examples/01-blog/Blog/Type/StoryType.php [hunk 12]
```diff
@@ -21,9 +20,12 @@ public function __construct()
             'name' => 'Story',
             'fields' => static fn (): array => [
                 'id' => Types::id(),
-                'author' => Types::user(),
-                'mentions' => new ListOfType(Types::mention()),
-                'totalCommentCount' => Types::int(),
+                'author' => ['type' => Types::user(), 'resolve' => static fn (Story $story): ?User => DataSource::findUser($story->authorId)],
```
Comment (spawnia) [inline examples/01-blog/Blog/Type/StoryType.php:23]:
```suggestion
                'author' => [
                    'type' => Types::user(),
                    'resolve' => static fn (Story $story): ?User => DataSource::findUser($story->authorId),
                ],
```

File: examples/01-blog/Blog/Type/StoryType.php [hunk 13]
```diff
@@ -46,109 +55,39 @@ public function __construct()
...
                         ],
                     ],
+                    'resolve' => static fn (Story $story): array => DataSource::findLikes($story->id, 10),
                 ],
-                'likedBy' => new ListOfType(Types::user()),
-                'affordances' => new ListOfType(Types::storyAffordances()),
-                'hasViewerLiked' => Types::boolean(),
-
-                'body' => HtmlField::build('body'),
+                'likedBy' => ['type' => new ListOfType(Types::user()), 'resolve' => static fn (Story $story) => DataSource::findLikes($story->id, 10)],
```
Comment (spawnia) [inline examples/01-blog/Blog/Type/StoryType.php:60]:
```suggestion
                'likedBy' => [
                    'type' => new ListOfType(Types::user()),
                    'resolve' => static fn (Story $story) => DataSource::findLikes($story->id, 10),
                ],
```

File: examples/01-blog/Blog/Type/StoryType.php [hunk 14]
```diff
@@ -46,109 +55,39 @@ public function __construct()
...
+
+                        $isViewer = $context->viewer === DataSource::findUser($story->authorId);
+                        if ($isViewer) {
+                            $affordances[] = StoryAffordancesType::EDIT;
+                            $affordances[] = StoryAffordancesType::EDIT;
+                            $affordances[] = StoryAffordancesType::DELETE;
+                        }
+
+                        $isLiked = DataSource::isLikedBy($story->id, $context->viewer->id);
+                        $affordances[] = $isLiked
```
Comment (spawnia) [inline examples/01-blog/Blog/Type/StoryType.php:75]:
```suggestion
                        $affordances[] = DataSource::isLikedBy($story->id, $context->viewer->id)
```

**Review by spawnia — COMMENTED** [review]
Almost perfect, thank you

File: examples/01-blog/Blog/Type/CommentType.php [hunk 15]
```diff
@@ -19,8 +18,18 @@ public function __construct()
             'name' => 'Comment',
             'fields' => static fn (): array => [
                 'id' => Types::id(),
-                'author' => Types::user(),
-                'parent' => Types::comment(),
+                'author' => [
+                    'type' => Types::user(),
+                    'resolve' => static fn (Comment $comment): ?User => $comment->isAnonymous
+                            ? null
+                            : DataSource::findUser($comment->authorId),
```
Comment (spawnia) [inline examples/01-blog/Blog/Type/CommentType.php:25]:
```suggestion
                    'resolve' => static fn (Comment $comment): ?User => $comment->isAnonymous
                        ? null
                        : DataSource::findUser($comment->authorId),
```

File: examples/01-blog/Blog/Type/StoryType.php [hunk 16]
```diff
@@ -21,9 +20,15 @@ public function __construct()
...
-                'totalCommentCount' => Types::int(),
+                'author' => [
+                    'type' => Types::user(),
+                    'resolve' => static fn (Story $story): ?User => DataSource::findUser($story->authorId),
+                ],
+                'mentions' => [
+                    'type' => new ListOfType(Types::mention()),
+                    'resolve' => static fn (Story $story): array => DataSource::findStoryMentions($story->id),
+                ],
+                'totalCommentCount' => ['type' => Types::int(), 'resolve' => static fn (Story $story): int => DataSource::countComments($story->id)],
```
Comment (spawnia) [inline examples/01-blog/Blog/Type/StoryType.php:31]:
```suggestion
                'totalCommentCount' => [
                    'type' => Types::int(),
                    'resolve' => static fn (Story $story): int => DataSource::countComments($story->id),
                ],
```

File: examples/01-blog/Blog/Type/CommentType.php (hunk 11 above)
Comment (shmax) [reply examples/01-blog/Blog/Type/CommentType.php]: Nice one  b4bf939846a35e0487f824801e50ac4cce057541

**Comment by shmax** [comment]
Cool, thanks much for reviewing and helping out 👍

**Pull request merged**